### PR TITLE
Repair/Powerdown Mode: box-dragging to repair/powerdown units

### PR DESCRIPTION
**Pull request by dnqbob**
See #17339. Allow players to drag a box to repair, as well as originally to click one by one.

Make that little wrench more convenient and helpful to players.

Now with Powerdown (or building online/offline) mode with box-dragging.

clear highlight box

power number

### CODE REVIEW

**Comment by dnqbob** [comment]
Firstly, because of some mistakes I made I have to change the name of my branch for testers and some of our former discussions can be seen here: #17416

Latest:

Powerdown dragging will show you how many power will be saved/consumed. Help you manage power system more precisely.
power number

Bug fix: 

-Now you can fix your allies buildings, again. 
-Flash now with mutiple repair targets.

- Repair Selection Priority is modified due to that.
```
-- Default: buildings > combat vehicles and aircrafts > non-combat vehicles and aircrafts > allies buildings
-- Ctrl: all
-- Alt: combat vehicles and aircrafts > non-combat vehicles and aircrafts > buildings > allies buildings
```
New smart selection logic for buildings in Repair Mode and Powerdown Mode: for example if your selection-drag contains both affected and unaffected buildings, the unaffected buildings will turn into affected buildings while the others remains unchanged.
smart

New modifiers mode for Powerdown Mode: 
- Default: generally turn on/off with smart selection.
- Ctrl: Only turn off.
- Alt: Only turn on.
smart modifiers
(Default -> Alt -> Ctrl)



History:

Previously version with modifiers and virsual improvements and Powerdown Mode with box-dragging
> Repair Mode with modifiers and virsual improvements:
> modifier
> 
> Powerdown (or building online/offline) Mode with box-dragging.
> power down on

Previously version with virsual improvements, and highlight actors that can be repaired;
> Green box, and green repairable high light
> box color high light repairable
> 
> Repairable high light over selection high light:
> High light repairable
> 
> Currently no key modifiers for this function, but suggestions&discussions are welcome in review about this. As well as bug reports for the latest build.


Previously version with bug fixed and improvements.
> Cursor changing && dragging:
> Cursor Dragging


Previously(with Cursor bug). Thanks @tovl 's help to solve it :
> Cursor changing:
> Cursor
> 
> Drag/Click:
> dragging

**Comment by dnqbob** [comment]
@tovl

> but I would suggest to just make it affect buildings only. [...] better served eventually by a separate icon

I think I need a new cursor image to inform player for this,  but I still think dragging can be good at repair units, especially the fixed wings that circle in the sky.  In other CNC games, WW/EAP use very simple and straight forward way to fix this problem: when return to airport they get auto-fix. However,  in Openra we have to click on them when they are playing around with us cursor, and we have to make sure it is the very damaged one from a pile of aircrafts especially when some mods consider the fixed wings aircrafts don't have any collision size when idle. 

Meanwhile,  when at the interval of the battle,  drag repairing can save time for players to find repair pad for a group of units or click one by one to send. We can just hold our view on the conflict ahead at most of time. 

> This creates a messy interface which not only leads to cognitive overload, but each one of these we add also creates more edge cases

It is messy and cannot satisfy every requirements even go this far. I think for non-combat units,  the old WW way can be more helping: medic/mechanic use a fake weapon when ordered to attack enemy,  the weapon is just guard the very ground they stand and immediately get into work. I think in Openra we can use the same way but more detailedly,  like a fake weapon makes they auto follow the nearest combat unit in Attack Move/Normal Move. This will solve most of non-combat units selection/micromanage problem, meanwhile make non-combat units catch up with the attack team as well as not expose themselves at the very front. It is also the same as the original CNC in micromanaging, which means old players can be comfortable. 

That doesn't mean modifiers is a bad idea.  Sometimes we really want to select non-combat units immediately from the team. From these selection modifiers we can also perform some decent micromanaging.  The key is,  how to get used to it. Can players get used to it? How to get an beloved expert on promoting these functions to persuade players to get used to it? 

> I think what you have here for visuals is good enough for now (I especially like the colored unit decorators; nice touch!).

Thanks!  I am not really care about the mechanic of the color in fact. I just want the color fitting the mode theme (green for repair,  yellow/orange for power) and not get mixed up with the terrain color (especially the snow map) as well.  

As for visual language for orders,  especially the modifiers units selection,  I have an idea: using floating words on cursor or at the one side of selection rectangle, like:
Ctrl -> "Units first" 
Alt -> "Non-combat first" 
Normal -> "Combat Units first"
Straight forward,  messy though. 


> As for mod SDK vs standard mods: [...]
So should I post this in modSDK as well?  I hope my comments is detailed enough for modders/gamers to modify these functions I made.

**Comment by dnqbob** [comment]
@pchote 

> The feature has never worked well in OpenRA (it didn't work with ships for a long time, and it still doesn't work with aircraft), and I could get right behind removing it provided an alternative solution is found for D2k.

However I think it (Repair-cursor on units) works well in RA/CNC1, meanwhile in D2K repair units must send order to carry all is really annoying when units is close to repair pad. Just look at their path finding.

1. Just like what @dragunoff said, player rarely remembers where the service depot is. I always use this mode on units, especially ground/naval units. That is why I want to make it better, based on orignal function.

2. Aircraft is indeed somewhat inconvenient in repairing but it is not the fault of the repair mode, but the fault on the various aircrafts repairing problems in Openra. All of following problems make aircrafts on Openra engine uneasy to repair, players have to do more efforts which is unacceptable to get them properly repaired.

    2.1. I don't think only allow the first (even fully HP) aircraft on repair pad while making other load ammo from the airfield is a good logic for **group repairing order for aircrafts**. IMO at least, the aircraft group repair order should let all the dameged units wait in line to repair, while full HP units trying to load up if they need to. I use drag-repair solve part of this problem by only select damaged units (and highlight), but the effect by this problem is still blocking other aircrafts from repairing normally.

    2.2. As for **fixed wings aircrafts** that circle in the sky, dragging is the way better than clicking when players try to repair them. This problem becomes even more troublesome together with 2.1, which means click one by one is the only way to repair those birds and players have to catch up with them by using mouse clicking when they are circling. 

    2.3. For cnc1 mod it is the way a little strange to let helicopters have collision size when ordered to repair. The blocking by other helicopters is makes repair/load ammo system almost out work for air units.

3. Repair pad itself is inconvenient at first place, so not only latest CNC series but also some mods in RA2 platform abandand the orignal repair pad. Aircrafts get auto-repaired in airfield while loading ammos, vehicles get repaired in factory that produce them. Because RA/CNC1/D2K/TS are old games so we have to accept this legacy, but at least we can simplify the action that players take to get repaired. No needs to located a repair pad, no need to click more than 3 times.

> Dropping the selection modifiers and repair-cursor on units, and having this share input characteristics with the support powers (assuming someone can come up with a way that makes sense) would solve all of my concerns over enabling this in the default mods.

1. Different from @tovl, I think it is not selection modifiers are useless, it is the **selection priority** too strange for me to learn and use in game as a player. I think the very orignal selection priority (player units > player buildings) is just enough.

Just consider why a new player like me will use the modifiers in game. I don't care about **selection priority** or modifiers, or anything called game mechanic in the first place. I just want to use Alt key to select/find out if there is any non-combat units in team. It is OK if I get nothing from it, but it is not OK when I find all selected units are combat units due to selection priority in Openra. I will feel weird at first place and, if I am not a CNC RTS fans I will give up this game due to the management is "too confusing or complex". I, new player, focus on only non-combat units, please give me no exceptions on it.

In conlusion and IMO, there can be modifiers in default mods, just like:

Normal Selection:
--Default: select all units.
--Alt: **only** select non-combat units (or combat if you want).
--Shift: add unit to selection set.

Repair:
--Default: **only** buildings.
--Alt: **only** units.

Powerdown:
--Alt: **only** power on.

I tend to follow following rules when design:
    1.1 All function close to orignal game in default modifier, then I won't recieve complains about the game is to complex to new gamer. They can play it in old way if they like.
    1.2 Modifier is a sub mode, not a complex things called **different selection priority**.
    1.3 The less modifier the better. People have to remember the sub modes in 1.2, so make it simple to remember.
    
2. Repair-cursor on units is useful and, I think this function should not be deleted at least for non-default mods. This function need to be complete and extend rather than deleted, **after all it doesn't affect the default function of repair mode.**

3. Different from @tovl, I think global order buttons like repair and power down are still the general orders generator, so no needs to "share input characteristics with the support powers". (I probably misundertand both of your meanings here so I want to get the reason why it should "share input characteristics with the support powers")

**Comment by dnqbob** [comment]
@dragunoff 
> Adding a retreat/repair button on the command line could be a good relpacement.

@tovl
> that functionality might be better served eventually by a separate icon (possibly as part of a retreat button as proposed in #6167).

I think this function as a sub mode of repair mode can be more helpful and convenient, if you means retreat/repair button works like Company of Hero.

You have to first select units, then press the button. Then there is a problem: should all other full HP units also retreat to repairpad?

In CoH these units retreat with a buff in speed, especially for infantry units, but in Openra is not. Which means this retreat function does not work well and will expose the slow and fragile infantry units from the protection of meat sheild of tanks.

if only retreat damaged units, I think sub mode of repair mode will be the way better. Changing mode in Openra will not affect the current selection set, which means you don't have to first select then repair. I also make highlight boxs to inform players which units will go to repair pad when drag, make it more predictable and controllable before unpressing the mouse button.

**Comment by pchote** [comment]
Splitting discussion threads across multiple locations makes it difficult to follow, especially when looking back in the future. It would be best if we can these topics in #17416 until those discussions have run their course.

**Comment by dnqbob** [comment]
> Splitting discussion threads across multiple locations makes it difficult to follow, especially when looking back in the future. It would be best if we can these topics in #17416 until those discussions have run their course.
 OK 


> Splitting discussion threads across multiple locations makes it difficult to follow, especially when looking back in the future. It would be best if we can these topics in #17416 until the discussions have run their course.

OK dude. Only if I can change branch name without deleting the former branch in git.

**Comment by dnqbob** [comment]
Is there anyone find any bug in this branch?
@tovl

File: OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: This file is missing the header.

File: OpenRA.Mods.Common/Orders/SellOrderGenerator.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Orders/SellOrderGenerator.cs]: Same here.

File: OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs
Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: ok

File: OpenRA.Mods.Common/Orders/SellOrderGenerator.cs
Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/SellOrderGenerator.cs]: ok

File: OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs
Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: Strange here,  
note

note
I have already add this head in my code.

File: OpenRA.Mods.Common/Orders/SellOrderGenerator.cs
Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/SellOrderGenerator.cs]: ~~also strange here.~~

File: OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs
Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: During your test it is missing or something?

Comment (abcdefg30) [reply OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: No, I mean the copyright header.

Comment (dnqbob) [reply OpenRA.Mods.Common/Orders/PowerDownOrderGenerator.cs]: Oh! I don't notice that. I will fix it.

**Comment by dnqbob** [comment]
Any other bugs?

**Comment by dragunoff** [comment]
This seems like a useful feature and it works alright in-game however I do have some major concerns :thinking: 

The first one is that this introduces **two new selection mechanics** (repair and power on/off) each with it's own modifiers. This along with the already existing selection priorities and modifiers adds to the cognitive burden of players (which modifier does what in which mode). 

The second concern is that **box dragging is mainly (and traditionally) used for selection** (in other RTS games and in other software that has selection - file managers, graphics editing suites) so players are used to that. It's kinda wired in a player's brain that a box drag results in a selection. Breaking that convention doesn't seem like a good idea.

One might argue that this is a power user feature and I shouldn't be so concerned. But even then I think it's **not worth the maintenance cost** that adding this would induce. 

Apart from those concerns there is a visual bug - when drag-repairing only one of the actors flashed (the most recently built one). 

Also there is a crash when sending units to repair. To reproduce send a group of units to repair and when they start entering the service depot spam-drag repair orders around them. Exception log below:
<details>

```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object
  at OpenRA.Mods.Common.Activities.Move.Tick (OpenRA.Actor self) [0x0009c] in <7d9db1a879fb45a79e79611036b81d1c>:0 
  at OpenRA.Activities.Activity.TickOuter (OpenRA.Actor self) [0x00052] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Traits.ActivityUtils.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00015] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Activities.Activity.TickChild (OpenRA.Actor self) [0x00008] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Mods.Common.Activities.MoveAdjacentTo.Tick (OpenRA.Actor self) [0x00136] in <7d9db1a879fb45a79e79611036b81d1c>:0 
  at OpenRA.Activities.Activity.TickOuter (OpenRA.Actor self) [0x00083] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Traits.ActivityUtils.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00015] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Activities.Activity.TickChild (OpenRA.Actor self) [0x00008] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Activities.Activity.TickOuter (OpenRA.Actor self) [0x000ad] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Traits.ActivityUtils.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00015] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Actor.Tick () [0x0000f] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.World.Tick () [0x00123] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bc] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Game.LogicTick () [0x0003e] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Game.Loop () [0x000f1] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Game.Run () [0x0003c] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <e700d8dfa99f4dbfaa84114336b01482>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <e700d8dfa99f4dbfaa84114336b01482>:0
```

</details>

**Comment by PunkPun** [comment]
Additionaly this feature is pretty redundant for vehicles and aircraft. Vast majority of the time you only send the most wounded ones to heal up, you don't really want to retreat your tanks just for a new paintjob. So you'll be clicking individually.

The only real benifit I see is repairing after nukes and demo trucks as they damage lots of structures. But even then it's questionable. You having to individually click on stuff to repair is part of damage. Other than that you don't really need to repair stuff that often, a few units / structures a minute, often none. It really isn't that obnoxious, from my experience the playerbase considers it a good apm sink. So I'm not really sure if there are any worthwhile benefits to merging this pr

**Comment by dnqbob** [comment]
> Apart from those concerns there is a visual bug - when drag-repairing only one of the actors flashed (the most recently built one).

I quite don't understand this one. How to reproduce it? Is this problem(flash) on highlight box or something?

>Also there is a crash when sending units to repair. To reproduce send a group of units to repair and when they start entering the service depot spam-drag repair orders around them. 

I will look into this one. I just try 7 units on pad and this problem does not reproduce. I need more units to complete the test. And I have a feeling about this problem probably has something to do with the [crash bug](https://github.com/ABrandau/Shattered-Paradise-SDK/issues/27) I reported. Those two bug is almost the same except the upper class of moving order type. Maybe too many orders will cause this crash?

add: 30 units, the game also not crash but they will stuck at pad when I spam-drag repair orders. Can you give more information about the test case?

**Comment by dnqbob** [comment]
>The second concern is that box dragging is mainly (and traditionally) used for selection (in other RTS games and in other software that has selection - file managers, graphics editing suites) so players are used to that. It's kinda wired in a player's brain that a box drag results in a selection. Breaking that convention doesn't seem like a good idea.

There are also some games and painting software use box drag to directly delete things/create shapes within this area. Nothing to do with selection.

> Vast majority of the time you only send the most wounded ones to heal up, you don't really want to retreat your tanks just for a new paintjob. So you'll be clicking individually.

You can just click individually when you need. And there is also a chance you really need send all of the surviviors after a failure battle to repair pad directly, which I mean even not only RA mod for OpenRA that emphasis on formation.

As for aircrafts in all mods/games made by Openra engine, I think this function is at least better than the current clicking. Fixed wings aircrafts are always circling without collision size to each other, it is hard to click on damaged aircrafts when you have more than 5 of them piled with each other. And the system of current air repair is also not good, when directly send to pad they will send a random member to pad in ignorance of it damaged or not. Is this for balance to nerf aircrafts or just the thing that needs improving?

> You having to individually click on stuff to repair is part of damage.

This depends whether this factor will be a balance issue for mod/game made by Openra. Like maybe in some mod/game they get crazy idea about power consumming units that require this powerdown function and its modifiers, as well as individually click on stuff to balance is not considered in this game/mod by devs.

In addition, I think use this as a balance point is kinda reluctant. In other word since this one can be considered as a balance point, why won't we cancel all the addition order style inherited/inspired from cnc3 and other RTS,  from which this game becomes more convenient, to make player play fair on their own technique and APM?

>The only real benifit I see is repairing after nukes and demo trucks

In Openra RA mod, kinda correct. But during combat few buildings will stand with full HP, they will always receive some damage from aircraft crashes to miss fire.

Besides, for modders and game makers, you can not stop they from falling in love with support powers and massive damaging weapons. Just look at RV.

**Comment by dnqbob** [comment]
>One might argue that this is a power user feature and I shouldn't be so concerned. But even then I think it's not worth the maintenance cost that adding this would induce.

It depends on mod/game makers. The powerdown modifier just a idea to satisfy if anyone need frequently and precisly operation on powerdown function (for that crazy idea, again).  You can really ignore modifiers if you don't want this. (I can also simplified it to simple non-modifier version if you want) 

I do agree with less modifiers the better but I still need at least one modifiers on repair function. In addition,  the usage of even standard selection modifiers are still uncertain and in arguments,  so I am confused and want to wait until core devs make final decision about modifiers.  I will follow them if they need simplified the modifiers.

**Comment by dragunoff** [comment]
> > Apart from those concerns there is a visual bug - when drag-repairing only one of the actors flashed (the most recently built one).
> 
> I quite don't understand this one. How to reproduce it? Is this problem(flash) on highlight box or something?

When you target an individual unit or building (with attack, repair, capture, etc.) the whole sprite flashes briefly. With the box dragging only the last actor flashes. This may well be a limitation of the target flash code which was not designed for multiple actors. 

> add: 30 units, the game also not crash but they will stuck at pad when I spam-drag repair orders. Can you give more information about the test case?

It's reproducible even with a single unit. Send it to repair and keep repair-selecting it as it is getting near and entering the service depot. It is reproducible even by normal repair-clicking and it also happens on the bleed branch, so the culprit should be elsewhere :thinking:

**Comment by dnqbob** [comment]
> When you target an individual unit or building (with attack, repair, capture, etc.) the whole sprite flashes briefly. With the box dragging only the last actor flashes. This may well be a limitation of the target flash code which was not designed for multiple actors.

Thanks, I didn't notice this before your comment here. I will try to fix it.

>It's reproducible even with a single unit. Send it to repair and keep repair-selecting it as it is getting near and entering the service depot. It is reproducible even by normal repair-clicking and it also happens on the bleed branch, so the culprit should be elsewhere 🤔

I will also try look into this one (but I am not familiar with this part, so I guess this crash bug needs to be reported to @pchote).

**Comment by PunkPun** [comment]
> In addition, I think use this as a balance point is kinda reluctant. In other word since this one can be considered as a balance point, why won't we cancel all the addition order style inherited/inspired from cnc3 and other RTS, from which this game becomes more convenient, to make player play fair on their own technique and APM?

RTS games are all about managing your attention, deciding where you want to spend time clicking and looking. They differentiate from each other by what you need to pay attention to and what actions you need to do for minimal & optimal play. Some RTS's require minimal input and basically play themselves, on the other side of the spectrum they are manual with minimal to no automatic behaviour, some even have artificial mechanics that make it harder do basic actions like a 12 unit selection limit in Brood War. There are people who like games on the both sides of the spectrum, OpenRA is somewhere in the middle. 

The waypoint system let's you avoid some cognitive burden of constantly keeping track of individual unit actions to make them follow a path. People who play and those who developed OpenRA like it so it is kept in. Plus keeping track of all unit actions is still the optimal and higher level play. While in my experience this pr was met pretty negatively from the playerbase. It removes a simple and pretty enjoyable apm sink. 

> As for aircrafts in all mods/games made by Openra engine, I think this function is at least better than the current clicking. Fixed wings aircrafts are always circling without collision
size to each other, it is hard to click on damaged aircrafts when you have more than 5 of them piled with each other. And the system of current air repair is also not good, when directly send to pad they will send a random member to pad in ignorance of it damaged or not. Is this for balance to nerf aircrafts or just the thing that needs improving?

Clicking on planes with repair wrench is definitely annoying, it should be made better and simplifying it a single click would not generate resistance. This pr is definitely not the way to go as for all the issues listed above.

PS. You focus too much on modders. While they should always be considered, you'll never get a pr merged if it hurts the base mods

**Comment by PunkPun** [comment]
> And the system of current air repair is also not good, when directly send to pad they will send a random member to pad in ignorance of it damaged or not. Is this for balance to nerf aircrafts or just the thing that needs improving?

Not a random member, the first one that was ordered to repair. If depot is busy aircraft go to rearm instead. IMO they should go wait inline next to service depot but it's fine as it is

**Comment by dnqbob** [comment]
OK, I accept your idea on RTS. But I think sometimes we need to think out of the box a little bit.

>Clicking on planes with repair wrench is definitely annoying, it should be made better and simplifying it a single click would not generate resistance. This pr is definitely not the way to go as for all the issues listed above.
>Not a random member, the first one that was ordered to repair. If depot is busy aircraft go to rearm instead. IMO they should go wait inline next to service depot but it's fine as it is

So this is what I mentioned in the old branch. My method can only solve part of it (selecting damaged one). It is better to improve the entire aircraft repair system.



> PS. You focus too much on modders. While they should always be considered, you'll never get a pr merged if it hurts the base mods.

When I asked something related in the old branch, @pchote said:

>The changes would be merged into the engine, but disabled by default. Modders can then enable it if they choose in their mods.

Although I don't know in which way this can be shown to modder/game developer (by selecting in a graphic UI?), this one has no conflict with default mods if default mods don't want.

**Comment by dnqbob** [comment]
So if the default mod (RA) don't want this function for balance or something it can be set to disabled. But modders/game devs can use this if they want.

**Comment by dnqbob** [comment]
I think there is some misunderstanding here.  I am not going to change the default mods(RA, TD, D2K) and you can find the discussion we have at old branch link,  although I will be glad to see if one of them applies my idea. I am waiting for merging here is because **this pr can be merged as an alternative choice for modders and game maker**. So please don't worry about it. 

Just think about it,  some devs of RA mod has an idea about dropping the powerdown function because it now pretty useless at RA's mechanics. Why I have to bother myself to expand a function that may soon get disposaled?

**Comment by PunkPun** [comment]
> So if the default mod (RA) don't want this function for balance

it's not really because of balance, but it is a factor. It's more about that there are many people who dislike it.

> Just think about it, some devs of RA mod has an idea about dropping the powerdown function because it now pretty useless at RA's mechanics. Why I have to bother myself to expand a function that may soon get disposaled?

It may still take a while to happen if it will happen at all, balance changes go through a long community process to get merged and are usually organised by members of the community, not developers. They do however have the final say and take part in the process. For example last RA balance patch was by me. (I know that this section is irrelevant to this pr)

**Comment by dnqbob** [comment]
@dragunoff can I just ban the flash only in repair function? If I insist on fixing this problem then at least class `World` and many classes related will be changed, which may cause further chaos in codes. I would prefer changing only my classes, which can be more controllable.

**Comment by dnqbob** [comment]
~~Now I simply turn down the flash effect in repair orders generated by repair button, since highlight box can show the units need repairing, as well as the unexpected risks if I change the flash related classes.~~

**Comment by dnqbob** [comment]
@pchote Should I do something to this pr to make it more suitable for mod SDK choosing? Will it get merged into bleed or modders/game makers need to switch to my branch for this function?

**Review by Mailaender — DISMISSED** [review]
Works as promised.

**Comment by pchote** [comment]
To be honest, this PR has turned into a rather awkward and difficult situation, so I have been avoiding it. I apologise for this, its not fair to leave this hanging without comment.

It is never nice to reject a contribution (especially somebodies first), but this PR has grown a lot more complicated since it started, and I still have fundamental issues with the direction it has taken. I therefore can't support merging this, because I feel that the support and maintenance burden that this would place on me outweighs its utility for modders.

My main issue from the previous discussion, and which @dragunoff's [comment above](https://github.com/OpenRA/OpenRA/pull/17477#issuecomment-583093128) nicely summarizes, is that introducing a new set of selection behaviours is going to be confusing for players and introduce a non-trivial maintenance burden with a new set of edge cases and untested (because it won't be used in the default mods) code.

I also feel that removing the repair target flashing, even for the normal single-click repair, is a polish regression. Toggling many buildings at once also triggers #3844, which probably won't be fixable in the foreseeable future.

Quite a few modders use https://github.com/AttacqueSuperior/Engine as a base, it might be better to target this there instead?

**Comment by dnqbob** [comment]
> is that introducing a new set of selection behaviours is going to be confusing for players

I apply many modifiers here just because I don't know what is your suggestion on simplify this. As well as in the furture whether the selection modifiers and even selection priority will remain or not is still not decided by you. I will change if you inform me about these.

Besides, this function relys upon normal selection modifiers and selection priority when select from collection and by doing this I can leave the collection class unchanged. If you drop the most of normal selection modifiers and selection priority I will also have to change (but I already have the idea about the changing when it happens)

> and introduce a non-trivial maintenance burden with a new set of edge cases and untested (because it won't be used in the default mods) code.

How about use it in openRA TD? Since you have changed a lot of it and not many players care about TD for now, you can test this function also in TD. This one is looking convenient and make OpenRA have its own cool idea extend the orignal C&C, at least make it attractive in appearance in your demonstration video.

> I feel that the support and maintenance burden that this would place on me outweighs its utility for modders.

You can tell me on it and let me do it.

>Quite a few modders use https://github.com/AttacqueSuperior/Engine as a base, it might be better to target this there instead?

Ummm... thanks for this information. However, I personally believe OpenRA itself can be a good and simple 2D RTS engine not only for modders but also for game developers, although it is too early to say this at this stage. Therefore, if based on this thought, it is sometimes necessary to fullfill some of the function requirement even if it is not applied in the default games/mods for now. 

It is better to make a function intergration/intruduction of not only bleed branch engine itself but also all other modders/devs's branch engine, when your TS is done.

> I also feel that removing the repair target flashing, even for the normal single-click repair, is a polish regression. 

Orignal Powerdown function is not flash. Besides, I personllay hope you can fix the flash function and make it can flash on mutipile orders.

**Comment by dnqbob** [comment]
> I still have fundamental issues with the direction it has taken

Firstly, as for the repair global command can repair vehicle/aircraft, it is not very C&C but convenient, as well as it won't hurt the orignal C&C repair function (repair buildings) and somewhat sounds reasonable (the word of 'repair' in grammar can apply to buildings and machines noun). And it is also the point that I extend this repair function.

Secondarily, There are also some games and painting software use box drag to directly delete things/create shapes within this area. Nothing to do with adding to selection. **If you must dig deep in this problem, in popular RTS even clicking should be all selection on units, there should be no such thing as global order button/independent producing slot**, meanwhile all action(move, follow, repair, powerdown, skill, training and building) actors can perform is in the action slot UI for every selection collection. **First-select-then-act (If action is a vt., then choose a target)** is a popular common sense of RTS, even the common sense of all the softwares while most of C&C series are old weirdoes under this design theory and criticized by many players and modders(except General, which want to accept this idea and compeletly drop all former C&C UI elements). It will be better to add a action slot UI and cancel all the global order button/producing slot for OpenRA, and many intergration problem can be solved at the very beginning, if we look at RTS in this angle.

In conclusion, You can see this as an extention of orignal C&C repair function (it is wierd already) and your idea on vehicle repair, and all of these is fitting the word "repair" in grammar, fitting the image of a wrench in game functionally.

**Comment by dnqbob** [comment]
> I also feel that removing the repair target flashing, even for the normal single-click repair, is a polish regression. 

The cause of flash bug find, can be fixed easily here.

It is simply codes in `WorldInteractionControllerWidget` to avoid muti-flash on a series of orders passed, which is mainly designed for attack orders.

Now the flash works normaly and returns.

**Comment by dnqbob** [comment]
Becasue render and select interface are changing so even if I want to push here I need to wait for a stable version (now I made a switcher for this function but I think hardcore player will not accept this).

**Pull request closed without merging**
